parsel/selector.py

```diff
@@ -365,7 +398,9 @@ def xpath(
         try:
             xpathev = self.root.xpath
         except AttributeError:
-            return self.selectorlist_cls([])
+            return typing.cast(
+                SelectorList[_SelectorType], self.selectorlist_cls([])
```

Comment:
Do you think it can be solved by type annotating self.selectorlist_cls?

Reply:
In general, I think we should try to avoid typing.cast as much as possible. 
It could be ok to merge it as-is, as a temporary workaround.

Reply:
`self.selectorlist_cls` is already annotated to be `SelectorList["Selector"]`, but without this cast we get `Incompatible return value type (got "SelectorList[Selector]", expected "SelectorList[_SelectorType]")  [return-value]`. According to [the relevant test](https://github.com/scrapy/parsel/blob/deceb29371ae9b9bd2e2e11c2f8f2e6ba8f0e4df/tests/test_selector.py#L1023) the idea is to manually set this attribute to `class MySelectorList(SelectorList["MySelector"])` in `class MySelector(Selector)`, which of course is not a good design and I don't know how to express this in typing. This part also sounds like a good thing to replace with some modern metaprogramming just like SelectorList itself I mentioned earlier, or there may even exist a simple way to annotate `Selector` so that this field automatically has a correct type (something that always means `SelectorList[type(self)]` probably).

Reply:
> self.selectorlist_cls is already annotated to be SelectorList["Selector"]

Is it? I can see how it's assigned to `SelectorList["Selector"]`, but not annotated. Could we annotate it as `SelectorList[_SelectorType]`, and assign to `SelectorList["Selector"]`?

No sure it'd help, of course.

Reply:
`selectorlist_cls: Type[SelectorList[_SelectorType]] = SelectorList["Selector"]` leads to

```
parsel/selector.py:325: error: Type variable "parsel.selector._SelectorType" is unbound  [valid-type]
parsel/selector.py:325: note: (Hint: Use "Generic[_SelectorType]" or "Protocol[_SelectorType]" base class to bind "_SelectorType" inside a class)
```

And making `Selector` inherit from `Generic[_SelectorType]` (not sure if it even makes sense) gives "Need type annotation" errors for all places in test_selector.py for all `sel = Selector()` lines, and also gives "Incompatible types in assignment (expression has type "Type[SelectorList[Any]]", variable has type "Type[SelectorList[_SelectorType]]")" for the `selectorlist_cls` line.

Reply:
I think Generic is the way to go here, but as you can see, it might be tricky :) So, let's skip all this for now, and do the way you did it here.